addons/knowledge/controllers/main.py

```diff
@@ -0,0 +1,319 @@
+        The dictionary can also include a `reload_tree` entry that will signify the caller that the aside block
+        listing all articles should be reloaded. This can happen when the article moves from one section to another.
+        **Note**: The user needs "write" permission to remove another member from
+        the list. The user can always remove themselves from the list.
+        :param article_id: (int) article id
+        :param member_id: (int) member id
+        :param partner_id: (int) partner id. Typically given when the member is based on a parent article
+            and we need to desync the article from its parent (and ensure the partner to removed is well removed
+            after the desynchronization).
+        """
+        article = request.env['knowledge.article'].sudo().browse(article_id)
+        if not article.user_can_write:
+            return werkzeug.exceptions.Forbidden()
+        previous_category = article.category
+        if partner_id:  # If the permission is based on parent article -> Desync article
+            partner = request.env['res.partner'].browse(partner_id).exists()
+            article_member_partners_ids = article.article_member_ids.partner_id.ids
+            if not article or not partner or not article._desync_access_from_parents(article_member_partners_ids):
+                return {'success': False}
+            # remove the partner if was copied from parent
+            article.article_member_ids.filtered(lambda m: m.partner_id.id == partner_id).unlink()
+        elif not article.exists() or not article._remove_member(member_id):
+            return {'success': False}
+        result = {'success': True}
+        if article.category != previous_category:
+            result['reload_tree'] = True
+        return result
+
+    @http.route('/knowledge/article/set_internal_permission', type='json', auth='user')
+    def article_set_internal_permission(self, article_id, permission, **post):
+        """
+        Sets the internal permission of the given article.
+        The function returns a dictionary indicating whether the request succeeds (see: `success` key).
+        The dictionary can also include a `reload_tree` entry that will signify the caller that the aside block
+        listing all articles should be reloaded. This can happen when the article moves from one section to another.
+        **Note**: The user needs "write" permission to update the internal permission of the article.
+        :param article_id: (int) article id
+        :param permission: (string) permission
+        """
+        article = request.env['knowledge.article'].browse(article_id)
+        if not article.user_can_write:
+            return werkzeug.exceptions.Forbidden()
+        previous_category = article.category
+        if not article._set_internal_permission(permission):
+            return {'success': False}
```

Comment:
Not sure we need all those "success" stuff.

Not sure the _set_internal_permission method actually ever returns False.